yaxs/xaxs='o' option

lindsayplatt opened the pull request:
addresses #211 

gs <- gsplot()  %>%
    points(1:4, xlim=c(NA,10), xaxs='o')  %>%
    lines(1:6)
gs

Commit:
Merge branch 'master' of https://github.com/lindsaycarr/gsplot into lindsaycarr-master

Commit:
WIP

Commit:
changes

Commit:
added yaxs='o' feature for a single buffer

Commit:
buffer of 4% is based on the upper limit of the axis

lindsayplatt commented on R/calc_views.R:
```diff
@@ -135,6 +137,18 @@ set_view_lim <- function(views){
+      usr.axs <- data[[paste0(var, 'axs')]][[as.numeric(i)]]
+      view.index <- which(!names(views[[as.numeric(i)]]) %in% c('window', 'gsplot'))[which(!is.na(usr.axs))]
+      
+      if (any(!is.na(usr.axs)) && usr.axs[which(!is.na(usr.axs))] == 'o') {
```
@ldecicco-USGS @jread-usgs can you review this and help me make it less messy? There's so much indexing/subsetting craziness (lines 145-150 in particular)

jordansread replied:
I will take a look

jordansread commented on R/calc_views.R:
```diff
@@ -107,7 +107,9 @@ set_view_lim <- function(views){
   data <- list(y=summarize_args(views,c('y','y1','y0'),ignore=c('window','gs.config')), 
-               x=summarize_args(views,c('x','x1','x0'),ignore=c('window','gs.config')))
+               x=summarize_args(views,c('x','x1','x0'),ignore=c('window','gs.config')),
+               yaxs=summarize_args(views,c('yaxs'),ignore=c('window','gs.config')),
```
I would call `x` and `y` data, but the `yaxs` and `xaxs` probably shouldn't be called that (just for readability's sake)

lindsayplatt replied:
ah - ok, I'll put those in a separate list

jordansread replied:
I've got a PR coming to your branch. Another reason to avoid that is because you loop through all items in `data` on line 118, and you don't want to loop the `yaxs` right?

lindsayplatt replied:
It's actually not looping through the `yaxs`, because of line 117, which filters only the x and y. But I agree that it should be in a separate list anyways

jordansread commented:
also, just a note. I know it is confusing, but we are trying to be consistent w/ styles. So `camelCase` is to be avoided, even though there are tons of examples that use it within the package.

lindsayplatt commented:
Ok, I can't seem to find where I added that style of variable naming. Can you point it out? 

Also, very useful to know, I have been unsure about which is preferred.

jordansread commented on R/calc_views.R:
```diff
@@ -135,6 +137,18 @@ set_view_lim <- function(views){
+        buffer.lim <- ifelse(which(is.na(usr.lim)) == 1, buffer.lim - buffer, buffer.lim + buffer)
+        views[[as.numeric(i)]][['window']][[lim.name]][[which(is.na(usr.lim))]] <- buffer.lim
+        views[[as.numeric(i)]][[view.index]][[paste0(var, 'axs')]] <- NULL
+        views[['par']][[paste0(var, 'axs')]] <- 'i'
```
can you explain what is going on here? (setting par(xaxs) = 'i')?

lindsayplatt replied:
The `default.yaml` has `yaxs = 'r'` and `xaxs = 'r'`, so when the "o" style is selected we add the 4% buffer to one side and so will want the axes to have the 'i' style so that they have the exact limits we defined.

jordansread commented:
oh, I guess you didn't add any camelCase, sorry.

jordansread commented on R/calc_views.R:
```diff
@@ -135,6 +137,18 @@ set_view_lim <- function(views){
+      view.index <- which(!names(views[[as.numeric(i)]]) %in% c('window', 'gsplot'))[which(!is.na(usr.axs))]
+      
+      if (any(!is.na(usr.axs)) && usr.axs[which(!is.na(usr.axs))] == 'o') {
+        buffer <- 0.04*views[[as.numeric(i)]][['window']][[lim.name]][[2]] #4% buffer based on upper limit
```
does this work when the axes is reversed? How about when it is logged?

jordansread replied:
Also, I don't think this calculation mirrors the one done in R for buffers. you might want to take a look at that (I think it uses range or diff for the 4% and not just the top lim)

lindsayplatt replied:
It does work for reversed axes because the reversal happens after the window limits are set:

``` r
gs <- gsplot()  %>%
    points(1:4, xlim=c(0,NA), xaxs='o')  %>%
    axis(side=1, reverse=TRUE)
gs
```

logged axes do not work (thanks for pointing that out)

jordansread replied:
ok, good to know. The logging is a challenge because it is handled within plot.window the way we have `print` set up now

lindsayplatt replied:
You're right - it does use the range, it just so happened that on the example I used, it worked out with the upper limit

jordansread commented:
One way this can be cleaner is to separate the `user_lims` logic from the `data_lims`, perhaps as two separate functions that can both be tested against the different possible inputs/outputs.

lindsayplatt commented:
so two functions used within the `set_view_lim` function?

jordansread commented on R/calc_views.R:
```diff
@@ -135,6 +137,18 @@ set_view_lim <- function(views){
       views[[as.numeric(i)]][['window']][[lim.name]][!is.na(usr.lim)] <- usr.lim[!is.na(usr.lim)]
+      
+      usr.axs <- data[[paste0(var, 'axs')]][[as.numeric(i)]]
+      view.index <- which(!names(views[[as.numeric(i)]]) %in% c('window', 'gsplot'))[which(!is.na(usr.axs))]
```
should this be within the `if` block, since it is only relevant/used there?

lindsayplatt replied:
yes, another thing I missed

Commit:
suggestions for lcarr's PR

Commit:
merge fix

Merge branch 'lindsaycarr-master'

# Conflicts:
#	R/curve.R
#	man/curve.Rd

Commit:
yaxs/xaxs in separate list, view.index in "if" statement

jordansread commented:
There is also a nasty error when a user tries:

``` R
gs <- gsplot() %>%
    points(1:4, xlim=c(1,3), xaxs='o') %>%
    lines(1:6)
gs
```

lindsayplatt commented:
that should throw an error because you aren't indicating which limit should be buffered and which shouldn't

jordansread commented:
yeah, you might want to give it better text for the error to throw and check for it.

Commit:
pulling out buffer function

Commit:
Merge pull request #1 from jread-usgs/master

PR to lcarr's PR

Commit:
resolving conflicts

Merge branch 'master' of https://github.com/lindsaycarr/gsplot

Conflicts:
	R/calc_views.R
